opnode/l2/sync_start.go

```diff
@@ -0,0 +1,111 @@
+package l2
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/ethereum/go-ethereum"
+
+	"github.com/ethereum-optimism/optimistic-specs/opnode/eth"
+	"github.com/ethereum/go-ethereum/common"
+)
+
+var WrongChainErr = errors.New("wrong chain")
+
+func FindSyncStart(ctx context.Context, reference SyncReference, genesis *Genesis) (nextRefL1 eth.BlockID, refL2 eth.BlockID, err error) {
+	// 10 seconds for the whole thing (TODO: or do we want individual timeouts?)
+	ctx, cancel := context.WithTimeout(ctx, time.Second*10)
+	defer cancel()
+	var refL1 eth.BlockID
+	var parentL2 common.Hash
+	// Start at L2 head
+	refL1, refL2, parentL2, err = reference.RefByL2Num(ctx, nil, genesis)
+	if err != nil {
+		err = fmt.Errorf("failed to fetch L2 head: %v", err)
+		return
+	}
+	// Check if L1 source has the block
+	var currentL1 eth.BlockID
+	currentL1, _, err = reference.RefByL1Num(ctx, refL1.Number)
+	if err != nil {
+		if !errors.Is(err, ethereum.NotFound) {
+			err = fmt.Errorf("failed to lookup block %d in L1: %w", refL1.Number, err)
+			return
+		}
+		// if the L1 did not find the block, it might be out of sync.
+		// We cannot sync from L1 in this case, but must make sure we are not in
+		// a reorg to a L1 chain with less blocks.
```

Comment:
If the L2 is ahead in *height* it may still get reorged away, the canonical chain can be shorter and have more weight.

So we still need to traverse back the block numbers to resolve the reorg.